File: backend/apps/batches/selectors/execution.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any

from apps.batches.models import BatchStepStatus

if TYPE_CHECKING:
    from apps.batches.models import Batch, BatchStep


def get_batch_execution_payload(batch: Batch) -> dict[str, Any]:
    snapshot = batch.snapshot_json
    product = snapshot.get("product", {})
    steps = batch.steps.all()

    step_definitions = snapshot.get("steps", {})
    step_list: list[dict[str, Any]] = []
    current_step_id: int | None = None
    completed_count = 0
    applicable_count = 0

    for step in steps:
        step_def = step_definitions.get(step.step_key, {})
        sig_policy = step_def.get("signaturePolicy", {})

        step_list.append(
            {
                "id": step.id,
                "step_key": step.step_key,
                "sequence_order": step.sequence_order,
                "title": step.title,
                "kind": step_def.get("kind", ""),
                "status": step.status,
                "is_applicable": step.is_applicable,
                "signature_state": step.signature_state,
                "requires_signature": sig_policy.get("required", False),
            }
        )

        if step.is_applicable:
            applicable_count += 1
            if step.status in (BatchStepStatus.COMPLETE, BatchStepStatus.SIGNED):
                completed_count += 1
            elif current_step_id is None:
                current_step_id = step.id

    return {
        "id": batch.id,
        "batch_number": batch.batch_number,
        "status": batch.status,
        "product_name": product.get("productName", ""),
        "product_code": product.get("productCode", ""),
        "site": {
            "code": batch.site.code,
            "name": batch.site.name,
        },
        "template_name": snapshot.get("templateName", ""),
        "template_code": snapshot.get("templateCode", ""),
        "steps": step_list,
        "current_step_id": current_step_id,
        "progress": {
            "total": len(step_list),
            "completed": completed_count,
            "applicable": applicable_count,
        },
    }


def get_step_detail_payload(step: BatchStep) -> dict[str, Any]:
    batch = step.batch
    snapshot = batch.snapshot_json
    step_definitions = snapshot.get("steps", {})
    step_def = step_definitions.get(step.step_key, {})

    return {
        "id": step.id,
        "batch_id": batch.id,
        "step_key": step.step_key,
        "sequence_order": step.sequence_order,
        "title": step.title,
        "kind": step_def.get("kind", ""),
        "status": step.status,
        "is_applicable": step.is_applicable,
        "instructions": step_def.get("instructions", ""),
        "fields": step_def.get("fields", []),
        "signature_policy": step_def.get("signaturePolicy", {}),
        "blocking_policy": {
            "blocks_execution_progress": step.blocks_execution_progress,
            "blocks_step_completion": step.blocks_step_completion,
            "blocks_signature": step.blocks_signature,
            "blocks_pre_qa_handoff": step.blocks_pre_qa_handoff,
        },
        "data": step.data_json,
        "meta": step.meta_json,
    }

File: frontend/src/features/execution/schemas/execution.ts
import { z } from "zod";

export const siteBriefSchema = z.object({
  code: z.string(),
  name: z.string(),
});

export const stepSummarySchema = z.object({
  id: z.number(),
  step_key: z.string(),
  sequence_order: z.number(),
  title: z.string(),
  kind: z.string(),
  status: z.enum(["not_started", "in_progress", "complete", "signed"]),
  is_applicable: z.boolean(),
  signature_state: z.enum(["not_required", "required", "signed"]),
  requires_signature: z.boolean(),
});

export const progressSchema = z.object({
  total: z.number(),
  completed: z.number(),
  applicable: z.number(),
});

export const batchExecutionSchema = z.object({
  id: z.number(),
  batch_number: z.string(),
  status: z.string(),
  product_name: z.string(),
  product_code: z.string(),
  site: siteBriefSchema,
  template_name: z.string(),
  template_code: z.string(),
  steps: z.array(stepSummarySchema),
  current_step_id: z.number().nullable(),
  progress: progressSchema,
});

export const fieldDefinitionSchema = z.object({
  key: z.string(),
  type: z.string(),
  label: z.string(),
  required: z.boolean().optional(),
  options: z
    .array(z.object({ value: z.string(), label: z.string() }))
    .optional(),
  unit: z.string().optional(),
  validation: z.record(z.string(), z.unknown()).optional(),
});

export const signaturePolicySchema = z.object({
  required: z.boolean().default(false),
  meaning: z.string().default(""),
});

export const blockingPolicySchema = z.object({
  blocks_execution_progress: z.boolean(),
  blocks_step_completion: z.boolean(),
  blocks_signature: z.boolean(),
  blocks_pre_qa_handoff: z.boolean(),
});

export const stepDetailSchema = z.object({
  id: z.number(),
  batch_id: z.number(),
  step_key: z.string(),
  sequence_order: z.number(),
  title: z.string(),
  kind: z.string(),
  status: z.enum(["not_started", "in_progress", "complete", "signed"]),
  is_applicable: z.boolean(),
  instructions: z.string(),
  fields: z.array(fieldDefinitionSchema),
  signature_policy: signaturePolicySchema.default({ required: false, meaning: "" }),
  blocking_policy: blockingPolicySchema,
  data: z.record(z.string(), z.unknown()),
  meta: z.record(z.string(), z.unknown()),
});
